### PR TITLE
Supergraph coprocessor implementation

### PR DESCRIPTION
**Pull request by Geal**
*Description here*

Fixes #issue_number

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -28,6 +28,27 @@ Note that the name of this metric may change in the future.
...
+* label
+* data
+* path
+* errors
+* extensions
+
+When using `@defer` or subscriptions a supergraph response may contain multiple GraphQL responses, and the coprocessor will be called for each.
+
+By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/3408
+
```
Comment (Geal) [inline CHANGELOG.md:51]:
```suggestion
```

File: apollo-router/src/plugins/coprocessor/supergraph.rs [hunk 2]
```diff
@@ -0,0 +1,879 @@
...
+
+    let payload = Externalizable::supergraph_builder()
+        .stage(PipelineStep::SupergraphRequest)
+        .control(Control::default())
+        .and_id(TraceId::maybe_new().map(|id| id.to_string()))
+        .and_headers(headers_to_send)
+        .and_body(body_to_send)
+        .and_context(context_to_send)
+        .and_method(method)
+        .and_sdl(sdl_to_send)
```
Comment (o0Ignition0o) [inline apollo-router/src/plugins/coprocessor/supergraph.rs:202]: 🙌

File: .changesets/feat_geal_supergraph_coprocessor2.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+### Supergraph coprocessor implementation ([PR #3647](https://github.com/apollographql/router/pull/3647))
+
+This adds support for coprocessors at the supergraph service level. Supergraph plugins work on the request side with a parsed GraphQL request object, so the query and operation name, variables and extensions are directly accessible. On the response side, they handle GraphQL response objects, with label, data, path, errors, extensions. The supergraph response contains a stream of GraphQL responses, which can contain multiple elements if the query uses `@defer` or subscriptions. When configured to observe the responses, the coprocessor will be called for each of the deferred responses.
+
+By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/3647
```
Comment (Geal) [inline .changesets/feat_geal_supergraph_coprocessor2.md:5]:
```suggestion
### Supergraph coprocessor implementation ([PR #3647](https://github.com/apollographql/router/pull/3647))


Coprocessors now support supergraph service interception. 

On the request side, the coprocessor payload can contain:
- method
- headers
- body
- context
- sdl

On the response side, the payload can contain:
- status_code
- headers
- body
- context
- sdl

The supergraph request body contains:
* query
* operation name
* variables
* extensions

The supergraph response body contains:
* label
* data
* errors
* extensions

When using `@defer` or subscriptions a supergraph response may contain multiple GraphQL responses, and the coprocessor will be called for each.

By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/3647
```

**Review by BrynCooke — CHANGES_REQUESTED** [review]
Needs the metric added:
apollo.router.operations.coprocessor 

Plus update the test for apollo.router.config.coprocessor to ensure that it is doing the right thing.

**Comment by Geal** [comment]
> Needs the metric added:
> apollo.router.operations.coprocessor

[cb4d33b](https://github.com/apollographql/router/pull/3647/commits/cb4d33b5e209835ad1a49b9eaa1a6d6bb0149504)

> Plus update the test for apollo.router.config.coprocessor to ensure that it is doing the right thing.

[773bd02](https://github.com/apollographql/router/pull/3647/commits/773bd026b7c9e87817ba38799214103c84020ba0)

**Review by garypen — APPROVED** [review]

**Review by o0Ignition0o — APPROVED** [review]

**Review by BrynCooke — APPROVED** [review]

**Pull request merged**